### PR TITLE
IGNITE-16560 [Native Persistence 3.0] PageMemoryImpl porting

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-16560

### CODE REVIEW

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/ClockPageReplacementFlags.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+     * Constructor.
+     *
+     * @param totalPagesCnt Total pages count.
+     * @param memPtr Pointer to memory region.
+     */
+    ClockPageReplacementFlags(int totalPagesCnt, long memPtr) {
+        pagesCnt = totalPagesCnt;
+        flagsPtr = memPtr;
+
+        setMemory(flagsPtr, (totalPagesCnt + 7) >> 3, (byte) 0);
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/ClockPageReplacementFlags.java:50]: Please replace such places with GridUnsafe.zeroMemory(...)

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/GroupPartitionId.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        this.grpId = grpId;
+        this.partId = partId;
+    }
+
+    /**
+     * Returns flag to be used for partition.
+     *
+     * @param partId Partition ID.
+     */
+    public static byte getFlagByPartId(final int partId) {
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/GroupPartitionId.java:57]: I don't like this method, is it even used?

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/GroupPartitionId.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+    /**
+     * Returns page ID (64 bits) constructed from partition ID and given index.
+     *
+     * @param pageIdx Page Index, monotonically growing number within each partition.
+     */
+    private long createPageId(final int pageIdx) {
+        final int partId = getPartitionId();
+
+        return pageId(partId, getFlagByPartId(partId), pageIdx);
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/GroupPartitionId.java:141]: This looks weird as well

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/GroupPartitionId.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    /**
+     * Returns Full page ID. For index 0 will return super-page of next partition
+     *
+     * @param pageIdx Page Index, monotonically growing number within each partition
+     * @return FullPageId consists of cache ID (32 bits) and page ID (64 bits).
+     */
+    private FullPageId createFullPageId(final int pageIdx) {
+        return new FullPageId(createPageId(pageIdx), getGroupId());
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/GroupPartitionId.java:151]: If it's all just for meta page id, then we could just hardcode these expressions into "createFirstPageFullId", if it's even used

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryEx.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import org.apache.ignite.internal.configuration.storage.StorageException;
+import org.apache.ignite.internal.pagememory.PageMemory;
+import org.apache.ignite.internal.pagememory.metric.IoStatisticsHolder;
+import org.apache.ignite.lang.IgniteInternalCheckedException;
+
+/**
+ * Page memory with some persistence related additions.
+ */
+//TODO IGNITE-16350 Improve javadoc in this class.
+public interface PageMemoryEx extends PageMemory {
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryEx.java:30]: Is there a way to remove this interface completely? We should think about it later

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,1602 @@
...
+
+        return false;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long partitionMetaPageId(int grpId, int partId) {
+        assert started;
+
+        return pageId(partId, FLAG_DATA, 0);
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImpl.java:532]: I don't think that meta page has "data" flag. Does it? Please check

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,1602 @@
...
+
+            seg.acquirePage(absPtr);
+
+            if (!readPageFromStore) {
+                statHolder.trackLogicalRead(absPtr + PAGE_OVERHEAD);
+            }
+
+            return absPtr;
+        } catch (IgniteOutOfMemoryException oom) {
+            throw oom;
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImpl.java:707]: What is this? Please add a comment

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,1602 @@
...
+    /** {@inheritDoc} */
+    @Override
+    public int systemPageSize() {
+        return sysPageSize;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public int realPageSize(int grpId) {
+        return pageSize();
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImpl.java:751]: Please add TODO for the future

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 9]
```diff
@@ -252,6 +252,23 @@ public static String toHexString(byte[] arr, int maxLen) {
         return sb.toString().toUpperCase();
     }
 
+    /**
+     * Returns hex representation of memory region.
+     *
+     * @param addr Pointer in memory.
+     * @param len How much byte to read (should divide 8).
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:259]: Fix it.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/GroupPartitionId.java (hunk 2 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/GroupPartitionId.java:57]: Not used, remove the code associated with it.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/GroupPartitionId.java (hunk 4 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/GroupPartitionId.java:151]: The code is not needed got rid of it.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/ClockPageReplacementFlags.java (hunk 1 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/ClockPageReplacementFlags.java:50]: Added a todo

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryEx.java (hunk 5 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryEx.java:30]: Added a todo

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/ClockPageReplacementFlags.java (hunk 1 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/ClockPageReplacementFlags.java:50]: Fix it.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImpl.java (hunk 7 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImpl.java:707]: Forgot to remove when refactoring.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImpl.java (hunk 8 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImpl.java:751]: Already in **PageMemory#realPageSize**

**Review by ibessonov — APPROVED** [review]

**Pull request merged**